File: src/taichu/domain/models/writing_ai.py
"""Writing-page real AI run contracts."""

from __future__ import annotations

from enum import StrEnum
from typing import Any

from pydantic import Field

from taichu.domain.models.base import DomainModel


class WritingAIButtonType(StrEnum):
    """Writing-page AI buttons routed through the unified real LLM workflow."""

    CHAT = "chat"
    CONTINUE = "continue"
    POLISH = "polish"
    SETTING = "setting"
    SUGGESTION = "suggestion"
    EVIDENCE = "evidence"
    CHAPTER_SUMMARY = "chapter_summary"
    INSPIRATION = "inspiration"
    FACT = "fact"


class WritingAIReferenceScope(StrEnum):
    """Reference scope sent by the writing page."""

    NONE = "none"
    SELECTION = "selection"
    CHAPTER = "chapter"
    FULL_TEXT = "full_text"


class WritingAIRunStatus(StrEnum):
    """Synchronous run lifecycle checkpoints."""

    QUEUED = "queued"
    RETRIEVING = "retrieving"
    CALLING_LLM = "calling_llm"
    PARSING = "parsing"
    COMPLETED = "completed"
    FAILED = "failed"


class WritingAIOutputType(StrEnum):
    """Structured output type returned by the fixed prompt for each button."""

    CHAT_ANSWER = "chat_answer"
    TEXT_CANDIDATE = "text_candidate"
    POLISHED_TEXT = "polished_text"
    SETTING_SUGGESTION = "setting_suggestion"
    WRITING_SUGGESTION = "writing_suggestion"
    EVIDENCE_ANSWER = "evidence_answer"
    CHAPTER_SUMMARY = "chapter_summary"
    INSPIRATION = "inspiration"
    PENDING_FACT_CANDIDATES = "pending_fact_candidates"


class WritingAISelectionRange(DomainModel):
    """Optional selected text position supplied by the editor."""

    paragraph_start: int | None = None
    paragraph_end: int | None = None
    char_start: int | None = None
    char_end: int | None = None


class WritingAIInput(DomainModel):
    """Input snapshot saved for every writing AI run."""

    user_input: str = ""
    selected_text: str = ""
    selection_range: WritingAISelectionRange | None = None
    target_words: int | None = None
    draft_chapter_text: str | None = None


class WritingAIPromptSnapshot(DomainModel):
    """Prompt snapshot actually rendered for a writing AI run."""

    prompt_id: str = Field(min_length=1)
    prompt_version: str = Field(min_length=1)
    system_prompt: str = Field(min_length=1)
    user_prompt: str = Field(min_length=1)
    rendered_at: str = Field(min_length=1)


class WritingAIRetrievalEvidenceItem(DomainModel):
    """One chapter or knowledge evidence item used by the run."""

    item_id: str = Field(min_length=1)
    source_type: str = Field(min_length=1)
    source_id: str = Field(min_length=1)
    display_name: str = Field(min_length=1)
    excerpt: str = ""
    usage: str = ""


class WritingAIRetrievalContext(DomainModel):
    """Knowledge and evidence context injected into the prompt."""

    used: bool = True
    empty_reason: str | None = None
    items: list[WritingAIRetrievalEvidenceItem] = Field(default_factory=list)
    knowledge_context: str = ""
    evidence_context: str = ""


class WritingAIStructuredOutput(DomainModel):
    """Normalized model output stored after JSON parsing."""

    output_type: WritingAIOutputType
    content: dict[str, Any] = Field(default_factory=dict)


class WritingAIRun(DomainModel):
    """Persisted trace for one writing-page AI call."""

    run_id: str = Field(min_length=1)
    status: WritingAIRunStatus
    button_type: WritingAIButtonType
    button_label: str = Field(min_length=1)
    model: str = ""
    chapter_id: str = Field(min_length=1)
    chapter_title: str = ""
    reference_scope: WritingAIReferenceScope
    input: WritingAIInput
    prompt_snapshot: WritingAIPromptSnapshot | None = None
    retrieval_context: WritingAIRetrievalContext | None = None
    raw_llm_output: str = ""
    structured_output: WritingAIStructuredOutput | None = None
    error: str | None = None
    created_at: str = Field(min_length=1)
    updated_at: str = Field(min_length=1)

File: web/src/lib/types/writing-ai.ts
export type WritingAIButtonType =
  | "chat"
  | "continue"
  | "polish"
  | "setting"
  | "suggestion"
  | "evidence"
  | "chapter_summary"
  | "inspiration"
  | "fact";

export type WritingAIReferenceScope =
  | "none"
  | "selection"
  | "chapter"
  | "full_text";

export type WritingAIRunStatus =
  | "queued"
  | "retrieving"
  | "calling_llm"
  | "parsing"
  | "completed"
  | "failed";

export type WritingAIOutputType =
  | "chat_answer"
  | "text_candidate"
  | "polished_text"
  | "setting_suggestion"
  | "writing_suggestion"
  | "evidence_answer"
  | "chapter_summary"
  | "inspiration"
  | "pending_fact_candidates";

export type WritingAISelectionRange = {
  paragraph_start?: number | null;
  paragraph_end?: number | null;
  char_start?: number | null;
  char_end?: number | null;
};

export type WritingAIInput = {
  user_input: string;
  selected_text: string;
  selection_range?: WritingAISelectionRange | null;
  target_words?: number | null;
  draft_chapter_text?: string | null;
};

export type WritingAIPromptSnapshot = {
  prompt_id: string;
  prompt_version: string;
  system_prompt: string;
  user_prompt: string;
  rendered_at: string;
};

export type WritingAIRetrievalEvidenceItem = {
  item_id: string;
  source_type: string;
  source_id: string;
  display_name: string;
  excerpt: string;
  usage: string;
};

export type WritingAIRetrievalContext = {
  used: boolean;
  empty_reason?: string | null;
  items: WritingAIRetrievalEvidenceItem[];
  knowledge_context: string;
  evidence_context: string;
};

export type WritingAIStructuredOutput = {
  output_type: WritingAIOutputType;
  content: Record<string, unknown>;
};

export type WritingAIRun = {
  run_id: string;
  status: WritingAIRunStatus;
  button_type: WritingAIButtonType;
  button_label: string;
  model: string;
  model_id: string;
  model_display_name: string;
  upstream_model: string;
  wire_protocol: string;
  llm_call_id?: string | null;
  input_tokens?: number | null;
  cached_input_tokens?: number | null;
  output_tokens?: number | null;
  reasoning_tokens?: number | null;
  total_tokens?: number | null;
  cost_amount?: string | number | null;
  cost_currency: string;
  cost_kind: string;
  chapter_id: string;
  chapter_title: string;
  reference_scope: WritingAIReferenceScope;
  input: WritingAIInput;
  prompt_snapshot?: WritingAIPromptSnapshot | null;
  retrieval_context?: WritingAIRetrievalContext | null;
  raw_llm_output: string;
  structured_output?: WritingAIStructuredOutput | null;
  error?: string | null;
  created_at: string;
  updated_at: string;
};

export type CreateWritingAIRunRequest = {
  button_type: WritingAIButtonType;
  chapter_id: string;
  reference_scope: WritingAIReferenceScope;
  user_input?: string;
  selected_text?: string;
  selection_range?: WritingAISelectionRange | null;
  target_words?: number | null;
  draft_chapter_text?: string | null;
  model_id?: string;
};

export type WritingAIStreamEvent =
  | { type: "run_started"; run_id: string; model_id: string }
  | { type: "text_delta"; delta: string }
  | {
      type: "usage";
      input_tokens?: number | null;
      cached_input_tokens?: number | null;
      output_tokens?: number | null;
      reasoning_tokens?: number | null;
      total_tokens?: number | null;
    }
  | { type: "run_completed"; run_id: string; call_id?: string | null }
  | { type: "run_failed"; run_id: string; message: string };

export type WritingAIRunListResponse = {
  runs: WritingAIRun[];
  page: number;
  page_size: number;
  total: number;
};
